README.md
# Flutter Global Configuration

A flutter package for managing different configurations by merging them together and making
them available everythere inside the app via a singleton.

## Table of Contents

1. [Preamble](#preamble)
2. [Install](#install)
   * [pubspec.yaml](#pubspec.yaml)
3. [Import](#import)
4. [Loading configuration](#loading-configuration)
   * [Load from asset](#load-from-asset)
   * [Load from path](#load-from-path)
   * [Load from path into key](#load-from-path-into-key)
   * [Load from map inside .dart file](#load-from-map-inside-.dart-file)
   * [Load from url](#load-from-url)
5. [Using the configuration in your app](#using-the-configuration-in-your-app)
6. [Overriding configuration](#overriding-configuration)
7. [Full Example](#full-example)
8. [Changelog](#changelog)
9. [Copyright and license](#copyright-and-license)

## Preamble

This package is also part of the [EZ Flutter Framework](https://ez-flutter.de).

## Install

### pubspec.yaml

Update pubspec.yaml and add the following line to your dependencies.

```yaml
dependencies:
  global_configuration: ^1.6.0
```

Or use the nullsafety preview version.

```yaml
dependencies:
  basic_utils: ^2.0.0-nullsafety.1
```


## Import

Import the package with :

```dart
import 'package:global_configuration/global_configuration.dart';
```

## Loading configuration

There are many ways where you can store your configuration files and load them.

### Load from asset

#### Creating a configuration file inside assets/cfg folder

Create a .json file and place it under assets/cfg/
Example filename: assets/cfg/app_settings.json

```json
{
  "key1": "value1",
  "key2": "value2"
}
```

Remember to update your pubspec.yaml file, so the config can be loaded from the assets directory!
The config files must be placed under the assets/cfg/ directory.

```yaml
flutter:
 assets:
  - assets/cfg/
```

#### Load configuration from assets at app start

```dart
import 'package:flutter/material.dart';
import 'package:global_configuration/global_configuration.dart';

void main() async{
  await GlobalConfiguration().loadFromAsset("app_settings");
  runApp(MyApp());
}
class MyApp extends StatelessWidget {
  ...
}
```

### Load from path

#### Creating a configuration file anywhere

Create a file anywhere inside your app.

#### Load configuration from path at app start

```dart
import 'package:flutter/material.dart';
import 'package:global_configuration/global_configuration.dart';

void main(){
  await GlobalConfiguration().loadFromPath("/path/file.json");
  runApp(MyApp());
}
class MyApp extends StatelessWidget {
  ...
}
```

### Load from path into key

The same as loadFromPath, but the config is added under the given key.

```dart
import 'package:flutter/material.dart';
import 'package:global_configuration/global_configuration.dart';

void main(){
  await GlobalConfiguration().loadFromPathIntoKey("/path/file.json", "env_settings");
  runApp(MyApp());
}
class MyApp extends StatelessWidget {
  ...
}
```

### Load from map

#### Creating a configuration inside .dart file

Create a dart file which includes your configuration and import it in your main file.
Example filename: /config/app_settings.config.dart

```dart
final Map<String, String> appSettings = {
  "key1": "value1",
  "key2": "value2"
};
```

#### Load configuration from map at app start

```dart
import 'package:flutter/material.dart';
import 'package:global_configuration/global_configuration.dart';
import 'config/app_settings.config.dart';

void main(){
  GlobalConfiguration().loadFromMap(appSettings);
  runApp(MyApp());
}
class MyApp extends StatelessWidget {
  ...
}
```

### Load from url

It is possible to load any json configuration file from a url. Use the method loadFromUrl to load
the config via GET request.
Please consider using a try / catch. This method will throw an exception if the status code of the
GET request is not 200.

The method also accepts query parameters and headers! The header "Accept: application/json" is always
included.

#### Load configuration from url at app start

```dart
import 'package:flutter/material.dart';
import 'package:global_configuration/global_configuration.dart';

void main() async{
  try{
    await GlobalConfiguration().loadFromUrl("http://urlToMyConfig.com");
  }catch(e){
    // something went wrong while fetching the config from the url ... do something
  }
  runApp(MyApp());
}
class MyApp extends StatelessWidget {
  ...
}
```

## Using the configuration in your app

Instatiate the GlobalConfiguration class and call any get($key) method.

```dart
import 'package:flutter/material.dart';
import 'package:global_configuration/global_configuration.dart';

class CustomWidget extends StatelessWidget {

    CustomWiget(){
        // Access the config in the constructor
        print(GlobalConfiguration().getString("key1")); // prints value1
    }

    @override
     Widget build(BuildContext context) {
        // Access the config in the build method
        return Text(GlobalConfiguration().getString("key2")); // prints value2
     }
}
```

## Overriding configuration

Sometimes it is necessary to override data in the configuration. This can be done by the updateValue method.
This method makes a check for the given type of the value and then performs an update the configuration map.

```dart
  Map<String, String> appSettings = {
    "key1": "value1",
    "key2": 1
  };
  GlobalConfiguration().loadFromMap(appSettings); // loading the configuration
  print(GlobalConfiguration().getString("key1")); // output : value1
  print(GlobalConfiguration().getString("key2")); // output : 1
  GlobalConfiguration().updateValue("key1", "value2"); // update the value for key1
  GlobalConfiguration().updateValue("key2", 2); // update the value for key2
  print(GlobalConfiguration().getString("key1")); // output : value2
  print(GlobalConfiguration().getString("key2")); // output : 2
```

### Full example

You can find a full example in the [example folder](/example/main.dart).

## Changelog

For a detailed changelog, see the [CHANGELOG.md](CHANGELOG.md) file

## Copyright and license

MIT License

Copyright (c) 2019 Ephenodrom

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Changelog

## [2.0.0-nullsafety.1] - 2021-04-06

* Update dependencies

## [2.0.0-nullsafety.0] - 2021-03-05

* Migrate to nullsafety

## [1.6.0] - 2020-09-02

* Improve access to nested configuration (#16)
* Merged pull request #15

## [1.5.0] - 2020-01-29

* Add possability to update null values (#10)
* Improve config loading from file (#9)

## [1.4.0] - 2019-09-16

* Add new method updateValue(), that replaces setValue()
* Add new method addValue()
* Add new method clear()
* Add new Unit tests
* Update documentation

## [1.3.0] - 2019-08-20

* Add new method loadFromUrlIntoKey

## [1.2.0] - 2019-07-25

* Add new method loadFromPathIntoKey

## [1.1.0] - 2019-06-17

* Fix bug in setValue and improve usage of setValue method

## [1.0.0] - 2019-06-01

* Add new method loadFromPath

## [0.1.4] - 2018-12-17

* Add new method setValue

## [0.1.3] - 2018-12-09

* Add new method loadFromUrl

## [0.1.2] - 2018-11-26

* Improve README.md

## [0.1.1] - 2018-11-25

* Add loading file from asset folder

## [0.1.0] - 2018-11-25

* Add example folder

## [0.0.2] - 2018-11-25

* Fill README.md with content and examples

## [0.0.1] - 2018-11-25

* Initial release


test/global_configuration_test.dart
import 'dart:convert';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:global_configuration/global_configuration.dart';

void main() {
  test('Testing get value for key.', () {
    GlobalConfiguration().clear();
    final Map<String, String> config1 = {"key1": "value1", "key2": "value2"};

    final Map<String, String> config2 = {"key3": "value3", "key4": "value4"};
    GlobalConfiguration().loadFromMap(config1);
    GlobalConfiguration().loadFromMap(config2);
    expect(GlobalConfiguration().getString("key1"), "value1");
    expect(GlobalConfiguration().getString("key3"), "value3");
  });

  test('Testing get value for key.', () {
    GlobalConfiguration().clear();
    final Map<String, String> config1 = {"key1": "value1", "key2": "value2"};

    final Map<String, String> config2 = {"key3": "value3", "key4": "value4"};
    GlobalConfiguration().loadFromMap(config1);
    GlobalConfiguration().loadFromMap(config2);
    expect(GlobalConfiguration().getValue("key1"), "value1");
    expect(GlobalConfiguration().getValue("key3"), "value3");
  });

  test('Testing read a deep value', () async {
    GlobalConfiguration().clear();
    final jsonString = '{"values" : { "key1" : "value1", "key2" : "value2" } }';
    Map<String, dynamic> config1 = json.decode(jsonString);

    GlobalConfiguration().loadFromMap(config1);
    expect(GlobalConfiguration().getDeepValue<String>("values:key2"), "value2");
  });

  test('Testing read a color', () async {
    GlobalConfiguration().clear();
    final jsonString = '{"color" : "#F44336" }';
    Map<String, dynamic> config1 = json.decode(jsonString);

    GlobalConfiguration().loadFromMap(config1);
    expect(
        GlobalConfiguration().getDeepValue<Color>("color"), Color(0xFFF44336));
  });

  test('Testing updateValue().', () async {
    GlobalConfiguration().clear();
    final Map<String, dynamic> config1 = {"hello": "world", "foo": 123};
    GlobalConfiguration().loadFromMap(config1);

    GlobalConfiguration().updateValue("hello", "world!");
    expect(GlobalConfiguration().getString("hello"), "world!");
    GlobalConfiguration().updateValue("foo", 321);
    expect(GlobalConfiguration().getInt("foo"), 321);

    try {
      GlobalConfiguration().updateValue("foo", "321");
    } catch (e) {
      expect(GlobalConfiguration().getInt("foo"), 321);
      return;
    }
    fail("Expected Exception!");
  });

  test('Testing updateValue().', () async {
    GlobalConfiguration().clear();
    final Map<String, dynamic> config1 = {"hello": "world", "foo": 123};
    GlobalConfiguration().loadFromMap(config1);

    GlobalConfiguration().updateValue("hello", "world!");
    expect(GlobalConfiguration().getValue("hello"), "world!");
    GlobalConfiguration().updateValue("foo", 321);
    expect(GlobalConfiguration().getValue("foo"), 321);

    try {
      GlobalConfiguration().updateValue("foo", "321");
    } catch (e) {
      expect(GlobalConfiguration().getInt("foo"), 321);
      return;
    }
    fail("Expected Exception!");
  });

  test('Testing update null value.', () async {
    GlobalConfiguration().clear();
    final Map<String, dynamic> config1 = {"hello": "world", "foo": null};
    GlobalConfiguration().loadFromMap(config1);

    GlobalConfiguration().updateValue("foo", 321);
    expect(GlobalConfiguration().getInt("foo"), 321);
  });

  test('Testing update null value.', () async {
    GlobalConfiguration().clear();
    final Map<String, dynamic> config1 = {"hello": "world", "foo": null};
    GlobalConfiguration().loadFromMap(config1);

    GlobalConfiguration().updateValue("foo", 321);
    expect(GlobalConfiguration().getValue("foo"), 321);
  });

  test('Testing addValue', () async {
    GlobalConfiguration().clear();
    GlobalConfiguration().addValue("hello", "world");
    expect(GlobalConfiguration().getString("hello"), "world");
  });

  test('Testing addValue', () async {
    GlobalConfiguration().clear();
    GlobalConfiguration().addValue("hello", "world");
    expect(GlobalConfiguration().getValue("hello"), "world");
  });
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:global_configuration/global_configuration.dart';
import 'AppSettings.config.dart';
import 'DevSettings.config.dart';

void main() async {
  GlobalConfiguration().loadFromMap(appSettings).loadFromMap(devSettings);
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  MyApp() {
    // Access configuration at constructor
    GlobalConfiguration cfg = new GlobalConfiguration();
    print("Key1 has value ${cfg.getString("key1")}");
    print("Key2 has value ${GlobalConfiguration().getString("key2")}");
    print("Key5 has value ${cfg.getString("key5")}, this should be null!");
  }

  @override
  Widget build(BuildContext context) {
    // Access configuration at build method
    GlobalConfiguration cfg = new GlobalConfiguration();
    print("Key3 has value ${cfg.getString("key3")}");
    return MaterialApp(
      title: 'Welcome to Flutter',
      home: Scaffold(
        appBar: AppBar(
          title: Text('Welcome to Flutter'),
        ),
        body: Center(
          child: Text('Hello World'),
        ),
      ),
    );
  }
}


lib/global_configuration.dart
library global_configuration;

import 'dart:ui';

import 'package:flutter/services.dart' show rootBundle;
import 'package:http/http.dart' as http;
import 'dart:convert';

///
/// Class for managing different configuration.
///
/// Use it with GlobalConfiguration() to access the singleton.
///
class GlobalConfiguration {
  static GlobalConfiguration _singleton = GlobalConfiguration._internal();

  ///
  ///  Convert string hex color like #AABBCC to Color widget
  ///
  Color _hexStringToColor(String code) =>
      Color(int.parse(code.substring(1, 7), radix: 16) + 0xFF000000);

  factory GlobalConfiguration() {
    return _singleton;
  }

  GlobalConfiguration._internal();

  Map<String, dynamic> appConfig = Map<String, dynamic>();

  ///
  /// Loading a configuration [map] into the current app config.
  ///
  GlobalConfiguration loadFromMap(Map<String, dynamic> map) {
    appConfig.addAll(map);
    return _singleton;
  }

  ///
  /// Loading a json configuration file with the given [name] into the current app config.
  ///
  /// If the given [name] does not have the file extension .json, it will be automatically added.
  ///
  /// The file has to be placed at assets/cfg/.
  ///
  /// Use the **loadFromPath** method if you want to store the file at another place.
  ///
  Future<GlobalConfiguration> loadFromAsset(String name) async {
    if (!name.endsWith(".json")) {
      name = "$name.json";
    }
    String content = await rootBundle.loadString("assets/cfg/$name");
    Map<String, dynamic> configAsMap = json.decode(content);
    appConfig.addAll(configAsMap);
    return _singleton;
  }

  ///
  /// Loading a json configuration file from a custom [path] into the current app config.
  ///
  Future<GlobalConfiguration> loadFromPath(String path) async {
    String content = await rootBundle.loadString(path);
    Map<String, dynamic> configAsMap = json.decode(content);
    appConfig.addAll(configAsMap);
    return _singleton;
  }

  ///
  /// Loading a json configuration file from a custom [path] into the current app config with the given [key].
  ///
  Future<GlobalConfiguration> loadFromPathIntoKey(
      String path, String key) async {
    String content = await rootBundle.loadString(path);
    Map<String, dynamic> configAsMap = json.decode(content);
    appConfig.putIfAbsent(key, () => configAsMap);
    return _singleton;
  }

  ///
  /// Loading a configuration file from the given [url] into the current app config using
  /// a http GET request to fetch the configuration.
  ///
  /// The request can be modified with [queryParameters] and [headers].
  ///
  Future<GlobalConfiguration> loadFromUrl(String url,
      {Map<String, String>? queryParameters,
      Map<String, String>? headers}) async {
    Map<String, dynamic> configAsMap = await _getFromUrl(url,
        queryParameters: queryParameters, headers: headers);
    appConfig.addAll(configAsMap);
    return _singleton;
  }

  ///
  /// Loading a configuration file from the given [url] into the current app config with the given [key].
  ///
  /// It uses a http GET request to fetch the configuration.
  /// The request can be modified with [queryParameters] and [headers].
  ///
  Future<GlobalConfiguration> loadFromUrlIntoKey(String url, String key,
      {Map<String, String>? queryParameters,
      Map<String, String>? headers}) async {
    Map<String, dynamic> configAsMap = await _getFromUrl(url,
        queryParameters: queryParameters, headers: headers);
    appConfig.putIfAbsent(key, () => configAsMap);
    return _singleton;
  }

  ///
  /// Reads a value of any type from persistent storage for the given [key].
  ///
  dynamic get(String key) => appConfig[key];

  ///
  /// Reads a value of T type from persistent storage for the given [key].
  ///
  T getValue<T>(String key) => appConfig[key] as T;

  ///
  /// Reads a value of T type from persistent storage for the given [keyPath]
  /// when [keyPath] is a Json path separated by ':' like 'appColors:primaryColor'
  /// when our Json file is:
  ///
  /// ```
  /// {
  ///     "appColors": {
  ///         "primaryColor": "#2e7d32"
  ///     }
  /// }
  /// ```
  ///
  /// You can also use getDeepValue<Color> when the json color value is
  /// a string hexadecimal like "#2e7d32".
  ///
  T? getDeepValue<T>(String keyPath) {
    dynamic _value;

    keyPath.split(":").forEach((element) {
      if (_value == null)
        _value = appConfig[element];
      else
        _value = _value[element];
    });

    if (_value != null) {
      if (T == Color) _value = _hexStringToColor(_value);

      return _value as T;
    }

    return null;
  }

  ///
  /// Reads a [bool] value from persistent storage for the given [key], throwing an exception if it's not a bool.
  ///
  @Deprecated("Use getValue instead")
  bool getBool(String key) => appConfig[key];

  ///
  /// Reads a [int] value from persistent storage for the given [key], throwing an exception if it's not an int.
  ///
  @Deprecated("Use getValue instead")
  int getInt(String key) => appConfig[key];

  ///
  /// Reads a [double] value from persistent storage for the given [key], throwing an exception if it's not a double.
  ///
  @Deprecated("Use getValue instead")
  double getDouble(String key) => appConfig[key];

  ///
  /// Reads a [String] value from persistent storage for the given [key], throwing an exception if it's not a String.
  ///
  @Deprecated("Use getValue instead")
  String getString(String key) => appConfig[key];

  ///
  /// Clear the persistent storage. Only for Unit testing!
  ///
  void clear() => appConfig.clear();

  /// Write a value from persistent storage, throwing an exception if it's not
  /// the correct type
  @Deprecated("Use updateValue instead")
  void setValue(key, value) => value.runtimeType != appConfig[key].runtimeType
      ? throw ("wrong type")
      : appConfig.update(key, (dynamic) => value);

  ///
  /// Update the given [value] for the given [key] in the storage.
  ///
  /// The updated value is *NOT* persistent
  /// Throws an exception if the given [value] has not the same [Type].
  ///
  void updateValue(String key, dynamic value) {
    if (appConfig[key] != null &&
        value.runtimeType != appConfig[key].runtimeType) {
      throw ("The persistent type of ${appConfig[key].runtimeType} does not match the given type ${value.runtimeType}");
    }
    appConfig.update(key, (dynamic) => value);
  }

  ///
  /// Adds the given [value] at the given [key] to the storage.
  ///
  /// The key and value is *NOT* persistent
  ///
  void addValue(String key, dynamic value) =>
      appConfig.putIfAbsent(key, () => value);

  ///
  /// Adds the given [map] to the storage.
  ///
  add(Map<String, dynamic> map) => appConfig.addAll(map);

  ///
  /// Sends a HTTP GET request to the given [url] with the given [queryParameters] and [headers].
  ///
  Future<Map<String, dynamic>> _getFromUrl(String url,
      {Map<String, String>? queryParameters,
      Map<String, String>? headers}) async {
    String finalUrl = url;
    if (queryParameters != null) {
      queryParameters.forEach((k, v) {
        finalUrl += !finalUrl.endsWith("?") ? "?$k=$v" : "&$k=$v";
      });
    }
    if (headers == null) {
      headers = Map<String, String>();
    }
    headers.putIfAbsent("Accept", () => "application/json");
    var encodedUri = Uri.encodeFull(finalUrl);
    var response = await http.get(Uri.parse(encodedUri), headers: headers);
    if (response.statusCode != 200) {
      throw new Exception(
          'HTTP request failed, statusCode: ${response.statusCode}, $finalUrl');
    }
    return json.decode(response.body);
  }
}
